Fix block rewards failing to be inserted on archival heights; fixup DBs affected by this

Doy-lee opened the pull request:
It's possible for there to be duplicate rows on insertion when we populate the archival delayed payment rows. If we have a deregistration which lasts for 21k blocks which elapses over two consecutive 10k intervals, then, it will already exist in the archive.

Since we have a UNIQUE constraint on the `(block, tx_index, contrib_index)` and the row already exists, the insertion fails and the rewards are not correctly recorded. The fix here is to allow duplicates so we update the `INSERT INTO` command to `INSERT OR IGNORE INTO` to handle this situation.

We do however now have a network that cannot come to consensus on the correct amount of rewards to sign off on for claiming. The solution I've come up with is to hardcode the expected rewards value of an address that was affected in the 1'870'000th and 1'880'000th block. If the node's DB does not have those values, we know they are out of sync and issue a rescan by detach the blockchain to the next closest archive.

As of current with the fix applied on a DB that was affected by the failure on block 1'870'000, I was able to rederive the rewards value for the hardcoded address that other nodes that weren't affected by this which currently constitutes around~20% of the network. The other 80% are witnessing a version of the network which experienced a failure at 1'870'000 and/or 1'880'000. 

The majority of the network is currently then underpaying the designated Ethereum address's stake and so by fixing it we should be correctly paying out the right amount of SESH to those users.

Doy-lee committed:
On detach, require '< payout_height" instead of '<='

When we detach to a height, we're assigning the SQL DB to exactly,
'height'. The convention in block handling for the SQL DB is that when
the db is at 'height', all the delayed_payment rows with 'payout_height
== height' should already be processed and removed from the table.

In other words there should not be any delayed payment rows with
'payout_height == height' when we detach to the 'height'.

By having '<= payout_height' we're restoring a row to return a stake
that has already been returned. However, this should not affect stakes
because when we retrieve delayed payments on the next block, we look for
'height + 1' which will not satisfy the 'payout_height'. Then by
accepting the new block, we have a trigger which prunes the old delayed
payment rows from the database.

Initially this seemed like it'd cause a problem but is handled, however there
are some APIs that rely on and enumerate delayed payment rows. If that
API is called on detach, at this block where the delayed payment row is
incorrectly held onto for a block longer than it should, those
subsystems will report incorrect information.

One of the stakeholders of this API is the staking portal which relies
on this information to correctly forecast the state of stakes for
a given wallet address so fixing this has some knock-on benefits despite
core consensus not being affected by it.

Doy-lee committed:
Correctly ignore duplicate delayed payment rows in the archive

Doy-lee committed:
Add fixup to patch DBs w/ incorrect rewards value in retrospect

jagerman commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -665,7 +671,7 @@ void BlockchainSQLite::upgrade_schema() {
             -- Delayed payments
-            INSERT INTO delayed_payments_archive ({delayed_fields})
+            INSERT OR IGNORE INTO delayed_payments_archive ({delayed_fields})
                 SELECT {delayed_fields} FROM delayed_payments;
```
I don't think `INSERT OR IGNORE` is the right choice here because it's overly broad and will ignore errors of *any* of the table's constraints, but the intention here is only to silently ignore violations of the `(block_height, block_tx_index, contributor_index)` unique constraint.

E.g.:
```
sqlite> create table t (a int, b int, unique(a,b), check(a >= 0), check(b >= 0));
sqlite> insert or ignore into t values (1,2);
sqlite> insert or ignore into t values (1,2);
sqlite> insert or ignore into t values (1,-2);
sqlite> select * from t;
┌───┬───┐
│ a │ b │
├───┼───┤
│ 1 │ 2 │
└───┴───┘
```
where that last insert ought to be hitting a check constraint failure.

Instead `ON CONFLICT(...) DO NOTHING` lets you make it specific to just the specific conflict that should be ignored:
```
sqlite> insert into t values (1,2) on conflict(a, b) DO NOTHING;
sqlite> insert into t values (1,2) on conflict(a, b) DO NOTHING;
sqlite> insert into t values (1,-2) on conflict(a, b) DO NOTHING;
Runtime error: CHECK constraint failed: b >= 0 (19)
```

and so here:
```suggestion
            INSERT INTO delayed_payments_archive ({delayed_fields})
                SELECT {delayed_fields} FROM delayed_payments
                ON CONFLICT (block_height, block_tx_index, contributor_index) DO NOTHING;
```

jagerman commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -791,8 +797,9 @@ void BlockchainSQLite::blockchain_detached(
-                " WHERE ? BETWEEN height AND payout_height"_format(
+                " WHERE ? >= height AND ? < payout_height"_format(
                         "delayed_fields"_a = DELAYED_PAYMENTS_COLS, "suffix"_a = suffix),
+                static_cast<int64_t>(new_height),
```
SQLite indexed binding (?NNN) avoids needing to bind the same value twice:
```suggestion
                " WHERE ?1 >= height AND ?1 < payout_height"_format(
                        "delayed_fields"_a = DELAYED_PAYMENTS_COLS, "suffix"_a = suffix),
```

jagerman commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -975,6 +975,118 @@ bool Blockchain::init(
         }
     }
 
+    {
```
Could we move this chunk of code (at least the actual queries) into db_sqlite.cpp so that there's just one place to look for all possible queries?  I do realize the intention is eventually to remove this, but I'm thinking it might be a little while before we can (because people could have dormant service nodes that aren't active and don't get upgraded until months from now and reintroduce the problem if this code doesn't stick around for a while).

Doy-lee replied:
Yep, I have moved it into a member function in BlockchainSQLite

Doy-lee committed:
Use DO NOTHING clause to ensure constraint errors bubble up

Co-authored-by: Jason Rhinelander <jason@imaginary.ca>

Doy-lee committed:
Use indexed binding notation (?NNN) to DRY new_height binding

Co-authored-by: Jason Rhinelander <jason@imaginary.ca>

Doy-lee committed:
Suffix SQL SELECT on UPSERT w/ WHERE true to avoid parser ambiguity

Otherwise the parser when it attempts to execute the command will fail
due to parsing failures. See _Parsing Ambiguity_ at:
https://www3.sqlite.org/lang_UPSERT.html

> When the INSERT statement to which the UPSERT is attached takes its
> values from a SELECT statement, there is a potential parsing ambiguity.
> The parser might not be able to tell if the "ON" keyword is introducing
> the UPSERT or if it is the ON clause of a join. To work around this, the
> SELECT statement should always include a WHERE clause, even if that
> WHERE clause is just "WHERE true".
>
> Ambiguous use of ON:
>
> INSERT INTO t1 SELECT * FROM t2
> ON CONFLICT(x) DO UPDATE SET y=excluded.y;
> Ambiguity resolved using a WHERE clause:
>
> INSERT INTO t1 SELECT * FROM t2 WHERE true
> ON CONFLICT(x) DO UPDATE SET y=excluded.y;

Doy-lee committed:
Add SQL rewards fixup for 1890k and table drive fixup via array

Doy-lee committed:
Sort records by oldest to rescan from newest height

Doy-lee committed:
Move SQL fixup into BlockchainSQLite function

Doy-lee committed:
Reverse the fixup order, add some post-failure checks, merge rescan checks
